service/history/transferQueueActiveTaskExecutor.go

```diff
@@ -273,17 +274,18 @@ func (t *transferQueueActiveTaskExecutor) processCloseExecution(
 	workflowCloseTime := wfCloseTime
 	workflowStatus := executionState.Status
 	workflowHistoryLength := mutableState.GetNextEventID() - 1
+	taskQueue := executionInfo.TaskQueue
 
 	startEvent, err := mutableState.GetStartEvent()
 	if err != nil {
 		return err
 	}
 	workflowStartTime := timestamp.TimeValue(startEvent.GetEventTime())
 	workflowExecutionTime := getWorkflowExecutionTime(mutableState, startEvent)
-	visibilityMemo := getWorkflowMemo(executionInfo.Memo)
-	searchAttr := getSearchAttributes(executionInfo.SearchAttributes)
+	visibilityMemo := getWorkflowMemo(copyMemo(executionInfo.Memo))
+	searchAttr := getSearchAttributes(copySearchAttributes(executionInfo.SearchAttributes))
```

Comment:
I agree that search attributes needs to be copied since they can be changed with special command but `memo` and `signalInfo.RequestId` are immutable. What is the practical reason of coping them? For consistency, though, I agree it is better to copy them in case if we add command to modify memo in future.

Reply:
one never know what functionality will be implemented in the future. 
if we make special case here, i.e. assuming some data structure will not be touched, and use after lock release, then in the future, if the behavior got changed, weird issue can arise.